controllers/drain_controller.go

```diff
@@ -0,0 +1,127 @@
+)
+
+type DrainReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=sriovnetwork.openshift.io,resources=sriovoperatorconfigs,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (dr *DrainReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	req.Namespace = namespace
+	reqLogger := log.FromContext(ctx).WithValues("drain", req.NamespacedName)
+	reqLogger.Info("Reconciling Drain")
+
+	config := &sriovnetworkv1.SriovOperatorConfig{}
+	err := dr.Get(ctx, types.NamespacedName{
+		Name: constants.DefaultConfigName, Namespace: namespace}, config)
+	if err != nil {
+		reqLogger.Error(err, "Error occurred on GET SriovOperatorConfig request from API server.")
+		return reconcile.Result{}, err
+	}
+
+	nodeList := &corev1.NodeList{}
+	err = dr.List(ctx, nodeList)
+	if err != nil {
+		// Failed to get node list
+		reqLogger.Error(err, "Error occurred on LIST nodes request from API server")
+		return reconcile.Result{}, err
+	}
+
+	// sort nodeList to iterate in the same order each reconcile loop
+	sort.Slice(nodeList.Items, func(i, j int) bool {
+		return strings.Compare(nodeList.Items[i].Name, nodeList.Items[j].Name) == -1
+	})
+
+	reqLogger.Info("Max node allowed to be draining at the same time", "MaxParallelNodeConfiguration", config.Spec.MaxParallelNodeConfiguration)
+
+	drainingNodes := 0
+	for _, node := range nodeList.Items {
+		if utils.NodeHasAnnotation(node, constants.NodeDrainAnnotation, constants.AnnoDraining) || utils.NodeHasAnnotation(node, constants.NodeDrainAnnotation, constants.AnnoMcpPaused) {
+			drainingNodes++
+		}
+	}
+
+	reqLogger.Info("Count of draining", "drainingNodes", drainingNodes)
+	if config.Spec.MaxParallelNodeConfiguration != 0 && drainingNodes == config.Spec.MaxParallelNodeConfiguration {
+		reqLogger.Info("MaxParallelNodeConfiguration limit reached for draining nodes")
+		return reconcile.Result{}, nil
+	}
+
+	for _, node := range nodeList.Items {
+		if utils.NodeHasAnnotation(node, constants.NodeDrainAnnotation, constants.AnnoDrainRequired) {
+			if config.Spec.MaxParallelNodeConfiguration == 0 || drainingNodes < config.Spec.MaxParallelNodeConfiguration {
+				reqLogger.Info("Start draining node", "node", node.Name)
+				patch := []byte(fmt.Sprintf(`{"metadata":{"annotations":{"%s":"%s"}}}`, constants.NodeDrainAnnotation, constants.AnnoDraining))
+				err = dr.Client.Patch(context.TODO(), &node, client.RawPatch(types.StrategicMergePatchType, patch))
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+				drainingNodes++
+			} else {
+				return reconcile.Result{}, nil
```

Comment:
Can you add a log that says there are too many draining nodes for the current configuration? I think it would help when debugging